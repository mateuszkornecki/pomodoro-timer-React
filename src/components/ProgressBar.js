import React from "react";
import classNames from "classnames";

function ProgressBar({ className = "", percent, trackRemaining, barColor }) {
    let progressClassName = classNames(
        "progress",
        className,
        { "progress--reverse": trackRemaining }
    );

    function checkColor(colorName) {
        if (barColor === colorName) return true
    }

    let progressBarClassName = classNames(
        "progress__bar",
        {
            "progress__bar--red": checkColor("red"),
            "progress__bar--blue": checkColor("blue"),
            "progress__bar--green": checkColor("green")
        }
    )

    const ProgressBarWidth =
        trackRemaining === true ? `calc(100 % - ${percent} %)` : `${percent}% `;

    return (
        <div className={progressClassName}>
            <div className={progressBarClassName} style={{ width: `${ProgressBarWidth} ` }}></div>
        </div>
    );
}

function isNumberBetweenRange(props, propName, componentName) {
    if (typeof props[propName] !== "number") {
        return new Error(`${propName} has to be number!`);
    } else if (parseInt(props[propName], 10) > 100 || parseInt(props[propName], 10) < 0) {
        return new Error(`An error in ${componentName}, ${propName} has to be between 10 and 50 and now it is equal to ${props[propName]}`);
    }
}

ProgressBar.propTypes = {
    percent: isNumberBetweenRange
}

export default ProgressBar;
