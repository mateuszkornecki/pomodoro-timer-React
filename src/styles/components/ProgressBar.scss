.ProgressBar {
    height: 30px;
    width: 100%;
    border: 1px solid white;
    margin-bottom: 1rem;
    display: flex;
    align-items: center;
    > div {
        /* width: calc(100% - 25%); */
        height: 20px;
        margin: auto 4px;
        background: #5c5c5c;
    }
}
