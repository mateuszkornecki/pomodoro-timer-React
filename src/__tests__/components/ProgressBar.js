import React from 'react';
import ReactDOM from 'react-dom';
import ProgressBar from '../../components/ProgressBar';
import TestRenderer from 'react-test-renderer'; // ES6


var root = null;
var progressRenderer = null;

beforeEach(() => {
    root = document.createElement("div");
    ReactDOM.render(
        <ProgressBar percent={10} barColor="blue" />
        , root)
});

describe('ProgressBar', () => {
    describe('tests for rendering', () => {
        it('render div when given certain percent and barColor', () => {
            expect(root.childNodes[0].nodeName).toBe("DIV");
        });
        it('has className ="progress" when given certain percent and barColor ', () => {
            expect(root.childNodes[0].className).toBe("progress");
        });
        it('has className = "progress__bar--blue" when given barColor="blue" ', () => {
            expect(root.childNodes[0].childNodes[0].className).toMatch(/progress__bar--blue/);
        });
        it('has width = "10%" when given percent={10} ', () => {
            expect(root.childNodes[0].childNodes[0].style.width).toBe("10%");
        });
    });
});

beforeEach(() => {
    progressRenderer = TestRenderer.create(
        <ProgressBar percent={10} barColor="blue" />
    );
});
describe('ProgressBar with react-test-renderer', () => {
    describe('when given percent and barColor', () => {
        it('render properly', () => {
            expect(progressRenderer.toJSON()).toMatchSnapshot();
        });
        it('renders div when given certain percent and barColor', () => {
            expect(progressRenderer.toJSON().type).toBe("div");
        });
        it('has className ="progress" when given certain percent and barColor ', () => {
            expect(progressRenderer.toJSON().props.className).toBe("progress");
        });
        it('has className = "progress__bar--blue" when given barColor="blue" ', () => {
            expect(progressRenderer.toJSON().children[0].props.className).toMatch(/progress__bar--blue/);
        });
        it('has width = "10%" when given percent={10} ', () => {
            expect(progressRenderer.toJSON().children[0].props.style.width).toMatch(/10%/);
        });
    });
});