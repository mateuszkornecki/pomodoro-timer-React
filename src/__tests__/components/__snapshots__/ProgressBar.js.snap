// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar with react-test-renderer when given percent and barColor render properly 1`] = `
<div
  className="progress"
>
  <div
    className="progress__bar progress__bar--blue"
    style={
      Object {
        "width": "10%  ",
      }
    }
  />
</div>
`;
